### ISSUE TITLE
Memory bug in trainer.py script

### ISSUE THREAD

**Issue by khosravani** [espnet/espnet#5921]
Hi,
While monitoring the memory allocation using NVIDA tools during ASR training, I encounter a bug in the espnet2/train/trainer.py. 
While doing a forward pass to calculate the loss in line 614 (retval = model(**batch)), the reference to the previous loss would cause the previous batch processing gradient to reside in memory (line 620). Since this is a significant memory allocation at least for E2E ASR, releasing this memory can free up to ~25% GPU memory.  With this bug it is difficult to tune the batch size without considering the memory used in the previous batch processing which is indead random. After this fix the memory is reset for a new batch processing, making it possible to increase the batch size and get the most out of the GPU memory for training speedup. In our case, this led to a 2x speedup.

This is a general bug which can lead to memory release during training and as a result training speedup by increasing the batch size.
 - OS information: Ubuntu 22.04.5 LTS
 - python version: Python 3.8.13
 - espnet version: latest
 - pytorch version  py3.8_cuda11.8_cudnn8.7.0_0
 -

**Comment by khosravani** [comment]
Before:
![Screenshot 2024-10-07 145640](https://github.com/user-attachments/assets/7296d832-c1d4-48a5-91ef-d0c01a68b879)

After:
![Screenshot 2024-10-07 145625](https://github.com/user-attachments/assets/c813fcd3-5165-4bbc-89cb-9c71cf91e55f)

**Comment by jctian98** [comment]
Thanks,
Could you share how you fix this issue? :)

**Comment by khosravani** [comment]
Your welcome, The fix is simply to release the reference to the loss variable as well as retval before line 614. The python garbage collector will release the memory.

**Comment by jctian98** [comment]
Thanks for the reply!
Sorry, maybe my knowledge is limited in this case.
If my understanding is correct, do you mean that: before line 614, the `loss`, and the corresponding computing graph of the last iteration still reside in the GPU memory? and we need to manually call `del loss` before line 614?

Since we already called the `loss.backward()` and `optimizer.step()` without `retain_graph=True` I was thinking the memory cache of the last iteration was already clear. Maybe I'm wrong.

**Comment by khosravani** [comment]
> Thanks for the reply! Sorry, maybe my knowledge is limited in this case. If my understanding is correct, do you mean that: before line 614, the `loss`, and the corresponding computing graph of the last iteration still reside in the GPU memory? and we need to manually call `del loss` before line 614?
> 
> Since we already called the `loss.backward()` and `optimizer.step()` without `retain_graph=True` I was thinking the memory cache of the last iteration was already clear. Maybe I'm wrong.

Anytime. Yes, as you see in the first picture, the big memory chunk (is proportional to B x T x U where U is the largest BPE sequence length) continues to reside in memory while the next iteration is doing the forward pass (right before it is assigned to retval variable). Setting loss and retval to None before this free up the space used by probably the gragh (see the second picture).

**Comment by khosravani** [comment]
Hi @sw005320 
I'm afraid the changes made to fix memory leak is not enough and still is there. 
First of all "del loss" will be ineffective when there is already another reference to the same memory location in retval["loss"].
Second I found out that an explicit call to torch.cuda.empty_cache() at the beggining of a new batch processing is necessary to resolve the memory leak. So I propose three lines of code as shown in the following picture.

![Screenshot 2024-10-21 173030](https://github.com/user-attachments/assets/febcdba7-8d3a-4295-b797-166e62b8e866)

**Comment by sw005320** [comment]
Thanks!
@wanchichen, can you check it?

**Comment by wanchichen** [comment]
Right, I missed the `retval`. However, I don't think we should do `torch.cuda.empty_cache()`, as each of those calls will be very expensive. 

Either way, running `torch.cuda.empty_cache()` should not actually affect our available memory, as it only clears memory that is already considered unallocated.

(See https://pytorch.org/docs/stable/generated/torch.cuda.memory_allocated.html)

**Comment by khosravani** [comment]
To give an explanation why torch.cuda.empty_cache() helps, here is the status of active cached segment timeline.

After a memory segment is freed by the application, instead of immediately returning the memory to the system, the CUDA runtime might cache it. This caching allows for faster subsequent allocations, as the memory can be reused without having to request new memory from the system.

In the following diagram, you see that overall a call to empty cache after processing each batch results in less GPU memory allocation at the end, helping me to increase the batch size even further. Howerver, I am not aware of the effect of this change on memory allocation speedup by caching. In case you need the files I can share it with you.

Before using torch.cuda.empty_cache()
![image](https://github.com/user-attachments/assets/61e9e091-6b18-4e3f-b205-fa25bbc00ed1)

After using torch.cuda.empty_cache()
![image](https://github.com/user-attachments/assets/07a673d5-48f1-4681-a775-6a4850d9f196)
